File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/service/User/UserServiceImpl.java
package com.ntnu.gidd.service.User;

import com.ntnu.gidd.dto.User.UserDto;
import com.ntnu.gidd.dto.User.UserPasswordResetDto;
import com.ntnu.gidd.dto.User.UserPasswordUpdateDto;
import com.ntnu.gidd.dto.User.UserRegistrationDto;
import com.ntnu.gidd.exception.*;
import com.ntnu.gidd.model.*;
import com.ntnu.gidd.repository.PasswordResetTokenRepository;
import com.ntnu.gidd.repository.PostRepository;
import com.ntnu.gidd.repository.TrainingLevelRepository;
import com.ntnu.gidd.repository.UserRepository;
import com.ntnu.gidd.service.Comment.CommentService;
import com.ntnu.gidd.service.Email.EmailService;
import com.ntnu.gidd.service.Registration.RegistrationService;
import com.ntnu.gidd.util.ContextAwareModelMapper;
import com.ntnu.gidd.service.post.PostService;
import com.ntnu.gidd.util.TrainingLevelEnum;
import com.querydsl.core.types.Predicate;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import javax.persistence.EntityNotFoundException;
import javax.transaction.Transactional;
import java.security.Principal;
import java.time.ZonedDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Implantation of a User Service
 */
@Slf4j
@NoArgsConstructor
@AllArgsConstructor
@Service
public class UserServiceImpl implements UserService {

	@Autowired
	private BCryptPasswordEncoder passwordEncoder;

	private ModelMapper modelMapper = new ContextAwareModelMapper();

	@Autowired
	private UserRepository userRepository;

	@Autowired
	private EmailService emailService;

	@Autowired
	private PasswordResetTokenRepository passwordResetTokenRepository;

	@Autowired
	private RegistrationService registrationService;

	@Autowired
	private CommentService commentService;

	@Autowired
	PostRepository postRepository;
	

	@Autowired
	private TrainingLevelRepository trainingLevelRepository;

	/**
	 * Method to get a user dto by email
	 * @param email
	 * @return the requested user
	 */
	@Override
	public UserDto getUserDtoByEmail(String email) {
		User user = userRepository.findByEmail(email).orElseThrow(UserNotFoundException::new);
		return modelMapper.map(user, UserDto.class);
	}

	/**
	 * Get user by email
	 * @param email
	 * @return requested user
	 */
	private User getUserByEmail(String email) {
		return userRepository.findByEmail(email).orElseThrow(UserNotFoundException::new);
	}

	/**
	 * Get user by id
	 * @param id
	 * @return requested user
	 */
	@Override
	public User getUserById(UUID id) {
		return userRepository.findById(id).orElseThrow(UserNotFoundException::new);
	}

	/**
	 * Helper method to check if a user with given email all ready exists
	 * @param email
	 * @return true/false
	 */
	private boolean emailExist(String email) {
		return userRepository.findByEmail(email).isPresent();
	}


	/**
	 * Method to create a user
	 * @param user
	 * @return User created
	 */
	@Override
	public UserDto saveUser(UserRegistrationDto user) {
		if (emailExist(user.getEmail())) {
			throw new EmailInUseException();
		}

		User userObj = modelMapper.map(user, User.class);
		userObj.setPassword(passwordEncoder.encode(user.getPassword()));
		return modelMapper.map(userRepository.save(userObj), UserDto.class);
	}

	/**
	 * Method to change passowrd on a user
	 * @param principal
	 * @param user
	 */
	@Override
	public void changePassword(Principal principal, UserPasswordUpdateDto user) {
		User userObj = getUserByEmail(principal.getName());
		userRepository.flush();
		if (passwordEncoder.matches(user.getOldPassword(), userObj.getPassword())) {
			userObj.setPassword(passwordEncoder.encode(user.getNewPassword()));
			userRepository.save(userObj);
		} else {
			throw new PasswordIsIncorrectException();
		}
	}

	/**
	 * Method to get all users
	 * @param predicate
	 * @param pageable
	 * @return list of users
	 */
	@Override
	public Page<UserDto> getAll(Predicate predicate, Pageable pageable) {
		return userRepository.findAll(predicate, pageable).map(user -> modelMapper.map(user, UserDto.class));
	}

	/**
	 * Method to get a user by id
	 * @param userId
	 * @return requested user
	 */
	@Override
	public UserDto getUserByUUID(UUID userId) {
		return modelMapper.map(userRepository.findById(userId)
												   .orElseThrow(UserNotFoundException::new), UserDto.class);
	}


	/**
	 * Method to deletes a user by email(username)
	 * @param username
	 * @return deleted user
	 */
	@Override
	public UserDto deleteUser(String username) {
		User user = userRepository.findByEmail(username).
				orElseThrow(UserNotFoundException::new);

		registrationService.deleteAllRegistrationsWithUsername(username);
		postRepository.deletePostsByCreator(user);
		commentService.deleteAllCommentsOnUser(username);
		userRepository.deleteById(user.getId());
		return modelMapper.map(user, UserDto.class);
	}

	/**
	 * Method to update a user by id
	 * @param id
	 * @param user
	 * @return Updated user info
	 */
	public UserDto updateUser(UUID id, UserDto user) {
		User updatedUser = userRepository.findById(id).orElseThrow(UserNotFoundException::new);
		updatedUser.setFirstName(user.getFirstName());
		updatedUser.setSurname(user.getSurname());
		updatedUser.setEmail(user.getEmail());
		updatedUser.setBirthDate(user.getBirthDate());
		updatedUser.setImage(user.getImage());
		if (user.getLevel() != null) updatedUser.setTrainingLevel(getTrainingLevel(user.getLevel()));
		return modelMapper.map(userRepository.save(updatedUser), UserDto.class);
	}
	
	private TrainingLevel getTrainingLevel(TrainingLevelEnum level) {
		return trainingLevelRepository.findTrainingLevelByLevel(level).
				orElseThrow(() -> new EntityNotFoundException("Training level does not exist"));
	}
	
	/**
	 * Finds a user, creates a token with the user and saves this.
	 * Sends a email to the user's email containing a link for resetting password
	 *
	 * @param email
	 * @return
	 */
	@SneakyThrows
	@Transactional
	public void forgotPassword(String email){
		User user = getUserByEmail(email);
		PasswordResetToken passwordResetToken = new PasswordResetToken();
		passwordResetToken.setUser(user);
		UUID id = passwordResetTokenRepository.save(passwordResetToken).getId();
		Map<String, Object> properties = new HashMap<>();
		properties.put("name", user.getFirstName() + " " + user.getSurname());
		properties.put("url", "https://gidd-idatt2106.web.app/auth/reset-password/" +id.toString() + "/");
		
		Mail mail = Mail.builder()
				.from("baregidd@gmail.com")
				.to(user.getEmail())
				.htmlTemplate(new HtmlTemplate("reset_password", properties))
				.subject("Reset password")
				.build();
		emailService.sendEmail(mail);
	}
	
	/**
	 * Finds a user and finds the token
	 * Checks if the token is valid and that the token is connected to the user
	 * If the checks pass, it sets a new password for the user
	 * Returns error if it doesn't pass the checks
	 *
	 * @param userReset
	 */
	public void validateResetPassword(UserPasswordResetDto userReset, UUID passwordResetTokenId) {
		User user = getUserByEmail(userReset.getEmail());
		
		PasswordResetToken passwordResetToken = passwordResetTokenRepository.findById(passwordResetTokenId).orElseThrow(ResetPasswordTokenNotFoundException::new);
		
		if (!user.equals(passwordResetToken.getUser()) && !validateResetToken(passwordResetToken)) {
			throw new InvalidResetPasswordToken();
		}
		user.setPassword(passwordEncoder.encode(userReset.getNewPassword()));
		userRepository.save(user);
	}
	
	private boolean validateResetToken(PasswordResetToken token) {
		return token.getExpirationDate().isBefore(ZonedDateTime.now());
	}
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/user/service/UserServiceImpl.kt
package ntnu.idatt2105.user.service

import com.querydsl.core.types.Predicate
import ntnu.idatt2105.core.exception.ApplicationException
import ntnu.idatt2105.core.exception.EntityType
import ntnu.idatt2105.core.exception.ExceptionType
import ntnu.idatt2105.core.mailer.HtmlTemplate
import ntnu.idatt2105.core.mailer.Mail
import ntnu.idatt2105.core.mailer.MailService
import ntnu.idatt2105.core.response.Response
import ntnu.idatt2105.core.util.CsvToBean.Companion.closeFileReader
import ntnu.idatt2105.core.util.CsvToBean.Companion.createCSVToBean
import ntnu.idatt2105.core.util.CsvToBean.Companion.throwIfFileEmpty
import ntnu.idatt2105.security.dto.ForgotPassword
import ntnu.idatt2105.security.dto.MakeAdminDto
import ntnu.idatt2105.security.dto.ResetPasswordDto
import ntnu.idatt2105.security.repository.PasswordResetTokenRepository
import ntnu.idatt2105.security.token.PasswordResetToken
import ntnu.idatt2105.security.token.isAfter
import ntnu.idatt2105.user.dto.DetailedUserDto
import ntnu.idatt2105.user.dto.UserDto
import ntnu.idatt2105.user.dto.UserRegistrationDto
import ntnu.idatt2105.user.model.RoleType
import ntnu.idatt2105.user.model.RoleType.USER
import ntnu.idatt2105.user.model.User
import ntnu.idatt2105.user.repository.RoleRepository
import ntnu.idatt2105.user.repository.UserRepository
import org.modelmapper.ModelMapper
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.BufferedReader
import java.io.InputStreamReader
import java.time.LocalDate
import java.util.*

@Service
class UserServiceImpl(
    val userRepository: UserRepository,
    val modelMapper: ModelMapper,
    val passwordEncoder: BCryptPasswordEncoder,
    val mailService: MailService,
    val passwordResetTokenRepository: PasswordResetTokenRepository,
    val roleRepository: RoleRepository
) : UserService {
    val logger = LoggerFactory.getLogger("UserServiceImpl")

    override fun registerUser(userDTO: UserRegistrationDto): UserDto {
        logger.info("Trying to register a user")
        if (existsByEmail(userDTO.email)) {
            logger.error("User already exists")
            throw ApplicationException.throwException(EntityType.USER, ExceptionType.DUPLICATE_ENTITY, "2", userDTO.email)
        }
        val user: User = createUserObj(userDTO)
        val savedUser: User = userRepository.saveAndFlush(user)
        forgotPassword(ForgotPassword(savedUser.email))
        logger.info("${user.id} has been created... Trying to send mail")
        return modelMapper.map(savedUser, UserDto::class.java)
    }

    private fun createUserObj(userDto: UserRegistrationDto): User {
        return User(id = UUID.randomUUID(), email = userDto.email, expirationDate = userDto.expirationDate, firstName = userDto.firstName,
            surname = userDto.surname, phoneNumber = userDto.phoneNumber)
    }

    override fun registerUserBatch(file: MultipartFile): Response {
        throwIfFileEmpty(file)
        logger.info("Trying to register multiple users")
        var fileReader: BufferedReader? = null

        try {
            fileReader = BufferedReader(InputStreamReader(file.inputStream))
            val csvToBean = createCSVToBean(fileReader, UserRegistrationDto::class.java)
            val listOfDTO: List<UserRegistrationDto> = csvToBean.parse()
            val listOfObj = mutableListOf<User>()
            if (listOfDTO.isEmpty()) throw Exception()
            listOfDTO.forEach {
                listOfObj.add(createUserObj(it))
            }
            userRepository.saveAll(listOfObj)
            logger.info("The users have been created. Sending emails...")
            listOfObj.forEach {
                forgotPassword(ForgotPassword(it.email))
            }
        } catch (ex: Exception) {
            throw throw ApplicationException.throwExceptionWithId(EntityType.USER, ExceptionType.NOT_VALID, "batch.invalidFile")
        } finally {
            closeFileReader(fileReader)
        }
        return Response("The users have been created")
    }

    private fun existsByEmail(email: String): Boolean {
        return userRepository.existsByEmail(email)
    }

    override fun getUsers(pageable: Pageable, predicate: Predicate): Page<UserDto> =
        userRepository.findAll(predicate, pageable).map { user -> modelMapper.map(user, UserDto::class.java) }

    override fun <T> getUser(id: UUID, mapTo: Class<T>): T {
        val user = getUserById(id)
        return modelMapper.map(user, mapTo)
    }

    override fun updateUser(id: UUID, user: UserDto): UserDto {
        val updatedUser = getUserById(id)
            .copy(
                firstName = user.firstName,
                surname = user.surname,
                email = user.email,
                phoneNumber = user.phoneNumber,
                image = user.image,
            )
        logger.info("${user.id} has been updated")
        return modelMapper.map(userRepository.save(updatedUser), UserDto::class.java)
    }

    override fun deleteUser(id: UUID): Response {
        val user = getUserById(id)
        userRepository.delete(user)
        logger.info("${user.id} has been deleted!")
        return Response("The user has been deleted")
    }

    override fun makeAdmin(user: MakeAdminDto): DetailedUserDto {
        val userObj = getUser(user.userId!!, User::class.java)
        val adminRole = roleRepository.findByName(RoleType.ADMIN)
        val userRole = roleRepository.findByName(USER)
        if (!userObj.roles.contains(adminRole))userObj.roles = mutableSetOf(adminRole!!, userRole!!)
        return modelMapper.map(userRepository.save(userObj), DetailedUserDto::class.java)
    }

    override fun invalidateExpiredUsers() {
        logger.debug("Invalidating expired users...")
        val invalidatedUsers = userRepository.findByExpirationDateBeforeAndRolesName(LocalDate.now(), USER)
            .map {
                it.roles.clear()
                logger.debug("Cleared roles of user with id: ${it.id}")
                it
            }
            .run {
                userRepository.saveAll(this)
            }
            .toList()

        logger.info("Successfully invalidated ${invalidatedUsers.size} users")
    }

    private fun getUserById(id: UUID): User =
        userRepository.findById(id).orElseThrow {
            throw ApplicationException.throwException(
                EntityType.USER,
                ExceptionType.ENTITY_NOT_FOUND,
                id.toString()
            )
        }

    private fun getUserByEmail(email: String): User =
        userRepository.findByEmail(email) ?: throw ApplicationException.throwException(
            EntityType.USER,
            ExceptionType.ENTITY_NOT_FOUND,
            email
        )

    override fun forgotPassword(email: ForgotPassword) {
        val user = getUserByEmail(email.email)
        val token = PasswordResetToken(user = user)
        token.id = passwordResetTokenRepository.save(token).id
        val properties = mapOf(
            1 to user.email,
            2 to "https://rombestilling.vercel.app/auth/reset-password/" + token.id + "/"
        )
        logger.info("Sending mail to ${user.id} for resetting password")
        sendEmail(user.email, properties)
    }

    override fun resetPassword(resetDto: ResetPasswordDto, id: UUID) {
        val token = passwordResetTokenRepository.findById(id).orElseThrow { throw ApplicationException.throwException(EntityType.TOKEN,
            ExceptionType.ENTITY_NOT_FOUND, id.toString())
        }
        val user = getUserByEmail(resetDto.email)
        if (user != token.user && token.isAfter()) {
            logger.error("$token is not valid for ${user.email}")
            throw ApplicationException.throwException(EntityType.TOKEN,
                ExceptionType.NOT_VALID, id.toString())
        }
        roleRepository.findByName(USER).run {
            if (this == null) throw ApplicationException.throwException(EntityType.ROLE,
                    ExceptionType.ENTITY_NOT_FOUND, this?.name!!)

            if (!user.roles.contains(this)) user.roles = mutableSetOf(this)
        }
        user.password = passwordEncoder.encode(resetDto.password)
        userRepository.save(user)
        logger.info("New password for ${user.email} has been created")
    }

    private fun sendEmail(email: String, properties: Map<Int, String>) {
        val mail =
            Mail("rombestilling@mail.com",
                properties.getValue(1),
                "Reset password",
                HtmlTemplate("reset_password", properties)
            )
        mailService.sendMail(mail)
    }

    override fun getReserverById(id: UUID) = getUserById(id)
}
